### ISSUE TITLE
Problems in multihead_attention

### ISSUE THREAD

**Issue by mikeymezher** [tensorflow/tensor2tensor#1441]
In T2T v1.12.0
Lines 3600-3643 in common_attention.py
`    if cache is None or memory_antecedent is None:
      q, k, v = compute_qkv(query_antecedent, memory_antecedent,
                            total_key_depth, total_value_depth, q_filter_width,
                            kv_filter_width, q_padding, kv_padding,
                            vars_3d_num_heads=vars_3d_num_heads)
    if cache is not None:
      if attention_type not in ["dot_product", "dot_product_relative"]:
        # TODO(petershaw): Support caching when using relative position
        # representations, i.e. "dot_product_relative" attention.
        raise NotImplementedError(
            "Caching is not guaranteed to work with attention types other than"
            " dot_product.")
      if bias is None:
        raise ValueError("Bias required for caching. See function docstring "
                         "for details.")

      if memory_antecedent is not None:
        # Encoder-Decoder Attention Cache
        q = compute_attention_component(query_antecedent, total_key_depth,
                                        q_filter_width, q_padding, "q",
                                        vars_3d_num_heads=vars_3d_num_heads)
        k = cache["k_encdec"]
        v = cache["v_encdec"]
      else:
        k = split_heads(k, num_heads)
        v = split_heads(v, num_heads)
        decode_loop_step = kwargs.get("decode_loop_step")
        if decode_loop_step is None:
          k = cache["k"] = tf.concat([cache["k"], k], axis=2)
          v = cache["v"] = tf.concat([cache["v"], v], axis=2)
        else:
          # Inplace update is required for inference on TPU.
          # Inplace_ops only supports inplace_update on the first dimension.
          # The performance of current implementation is better than updating
          # the tensor by adding the result of matmul(one_hot,
          # update_in_current_step)
          tmp_k = tf.transpose(cache["k"], perm=[2, 0, 1, 3])
          tmp_k = inplace_ops.alias_inplace_update(
              tmp_k, decode_loop_step, tf.squeeze(k, axis=2))
          k = cache["k"] = tf.transpose(tmp_k, perm=[1, 2, 0, 3])
          tmp_v = tf.transpose(cache["v"], perm=[2, 0, 1, 3])
          tmp_v = inplace_ops.alias_inplace_update(
              tmp_v, decode_loop_step, tf.squeeze(v, axis=2))
          v = cache["v"] = tf.transpose(tmp_v, perm=[1, 2, 0, 3])`

The block:
`      if memory_antecedent is not None:
        # Encoder-Decoder Attention Cache
        q = compute_attention_component(query_antecedent, total_key_depth,
                                        q_filter_width, q_padding, "q",
                                        vars_3d_num_heads=vars_3d_num_heads)
        k = cache["k_encdec"]
        v = cache["v_encdec"]`
should be unreachable, due to the sequence of preceding conditionals. 

Also - currently trying to debug why I'm having shape issues in dot product attention: 
`  with tf.variable_scope(
      name, default_name="dot_product_attention", values=[q, k, v]) as scope:

    q = tf.Print(q,[q],'ATTENTION QUERY: ')
    k = tf.Print(k,[k],'ATTENTION KEY: ')

    logits = tf.matmul(q, k, transpose_b=True)  # [..., length_q, length_kv]
    if bias is not None:
      bias = common_layers.cast_like(bias, logits)
      bias = tf.Print(bias,[bias],'BIAS: ')
      logits = tf.Print(logits,[logits],'LOGITS ')
      logits += bias
      logits = tf.Print(logits,[logits],'LOGITS POST ADD BIAS')

    # If logits are fp16, upcast before softmax
    logits = maybe_upcast(logits, activation_dtype, weight_dtype)

    weights = tf.nn.softmax(logits, name="attention_weights")

    weights = common_layers.cast_like(weights, q)
    weights = tf.Print(weights,[weights],'WEIGHTS POST CAST LIKE')

    if save_weights_to is not None:
      save_weights_to[scope.name] = weights
      save_weights_to[scope.name + "/logits"] = logits

    # Drop out attention links for each head.
    weights = common_layers.dropout_with_broadcast_dims(weights, 1.0 - dropout_rate, broadcast_dims=dropout_broadcast_dims)

    #Debugging weights last dimension not matching value first dimension
    weights = tf.Print(weights,[weights],'ATTENTION WEIGHTS AFTER DROPOUT: ')
    v = tf.Print(v,[v],'VALUE VECTOR')

    if common_layers.should_generate_summaries() and make_image_summary:
      attention_image_summary(weights, image_shapes)
    return tf.matmul(weights, v)
`

OUTPUT: 
BIAS: [[[[-0]]]]BIAS: [[[[-0]]]]
BIAS: [[[[-0]]]]
BIAS: [[[[-0]]]]
BIAS: [[[[-0]]]]
BIAS: [[[[-0]]]]

ATTENTION KEY: [[[[0.151335612 -0.404835284 1.54091191...]]]...]
VALUE VECTOR[[[[0.166511878 -1.3204602 -0.372632712...]]]...]
ATTENTION QUERY: [[[[0.0620518401 -0.0874474421 0.000517165521...]]]...]
LOGITS [[[[-0.137875706]][[1.05136979]][[-0.866865873]]]...]
LOGITS POST ADD BIAS[[[[-0.137875706]][[1.05136979]][[-0.866865873]]]...]
WEIGHTS POST CAST LIKE[[[[1]][[1]][[1]]]...]
ATTENTION WEIGHTS AFTER DROPOUT: [[[[1]][[1]][[1]]]...]
ATTENTION KEY: [[[[-0.493546933 0.678736925 1.12754095...]]]...]
VALUE VECTOR[[[[0.427185386 -0.0682374686 -0.551386356...]]]...]
ATTENTION QUERY: [[[[-0.290087074 0.0257696807 -0.136151105...]]]...]
LOGITS [[[[-0.889816523]][[0.497348607]][[-1.31549573]]]...]
LOGITS POST ADD BIAS[[[[-0.889816523]][[0.497348607]][[-1.31549573]]]...]
WEIGHTS POST CAST LIKE[[[[1]][[1]][[1]]]...]
ATTENTION WEIGHTS AFTER DROPOUT: [[[[1]][[1]][[1]]]...]
VALUE VECTOR[[[[0.394824564 -0.571578 -1.01061356...]]]...]
ATTENTION KEY: [[[[3.20773053 -1.05240285 0.20996505...]]]...]
ATTENTION QUERY: [[[[0.187075511 -0.0486161038 0.089396365...]]]...]
LOGITS [[[[1.59205556]][[-1.71665192]][[-1.37695456]]]...]
LOGITS POST ADD BIAS[[[[1.59205556]][[-1.71665192]][[-1.37695456]]]...]
WEIGHTS POST CAST LIKE[[[[1]][[1]][[1]]]...]
ATTENTION WEIGHTS AFTER DROPOUT: [[[[1]][[1]][[1]]]...]
ATTENTION QUERY: [[[[-0.136278272 -0.159313381 -0.204157695...]]]...]
VALUE VECTOR[[[[-1.74210846 1.38060951 -1.33350658...]]]...]
ATTENTION KEY: [[[[-0.646322787 -0.519498885 -0.108884774...]]]...]
LOGITS [[[[-1.05637074]][[-1.38687372]][[1.01201]]]...]
LOGITS POST ADD BIAS[[[[-1.05637074]][[-1.38687372]][[1.01201]]]...]
WEIGHTS POST CAST LIKE[[[[1]][[1]][[1]]]...]
ATTENTION WEIGHTS AFTER DROPOUT: [[[[1]][[1]][[1]]]...]
ATTENTION KEY: [[[[-0.766771 1.29248929 0.353579193...]]]...]
VALUE VECTOR[[[[-0.473533779 0.339176893 -0.285344601...]]]...]
ATTENTION QUERY: [[[[-0.00684091263 -0.0722586587 -0.0857348889...]]]...]
LOGITS [[[[-1.03410649]][[0.495673776]][[-1.93785417]]]...]
LOGITS POST ADD BIAS[[[[-1.03410649]][[0.495673776]][[-1.93785417]]]...]
WEIGHTS POST CAST LIKE[[[[1]][[1]][[1]]]...]
ATTENTION WEIGHTS AFTER DROPOUT: [[[[1]][[1]][[1]]]...]
ATTENTION KEY: [[[[-0.752419114 0.899734139 0.263004512...]]]...]
VALUE VECTOR[[[[-1.55072808 0.319361448 0.537948072...]]]...]
ATTENTION QUERY: [[[[-0.130305797 0.192893207 -0.0268632658...]]]...]
LOGITS [[[[-0.535128653]][[0.04286623]][[0.769883]]]...]
LOGITS POST ADD BIAS[[[[-0.535128653]][[0.04286623]][[0.769883]]]...]
WEIGHTS POST CAST LIKE[[[[1]][[1]][[1]]]...]
ATTENTION WEIGHTS AFTER DROPOUT: [[[[1]][[1]][[1]]]...]
BIAS: [[[[-0 -0]]]]
BIAS: [[[[-0 -0]]]]
BIAS: [[[[-0 -0]]]]
BIAS: [[[[-0 -0]]]]
BIAS: [[[[-0 -0]]]]
BIAS: [[[[-0 -0]]]]
ATTENTION KEY: [[[[-1.19346094 0.665090382 -0.502224...]]]...]
ATTENTION QUERY: [[[[0.193519413 -0.132038742 -0.0287191756...]]]...]
LOGITS [[[[-0.292171031]][[1.25045216]][[1.87848222]]]...]
VALUE VECTOR[[[[-0.436115 -0.63339144 0.0544230118...]]]...]
LOGITS POST ADD BIAS[[[[-0.292171031 -0.292171031]][[1.25045216...]]]...]
WEIGHTS POST CAST LIKE[[[[0.5 0.5]][[0.5...]]]...]
ATTENTION WEIGHTS AFTER DROPOUT: [[[[0.5 0.5]][[0.5...]]]...]

  File "/usr/local/lib/python2.7/dist-packages/tensor2tensor/layers/common_attention.py", line 3682, in multihead_attention
    activation_dtype=kwargs.get("activation_dtype"))
  File "/usr/local/lib/python2.7/dist-packages/tensor2tensor/layers/common_attention.py", line 1540, in dot_product_attention
    return tf.matmul(weights, v)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/math_ops.py", line 2019, in matmul
    a, b, adj_x=adjoint_a, adj_y=adjoint_b, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gen_math_ops.py", line 1245, in batch_mat_mul
    "BatchMatMul", x=x, y=y, adj_x=adj_x, adj_y=adj_y, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): In[0] mismatch In[1] shape: 2 vs. 1: [1,16,1,2] [1,16,1,64] 0 0
         [[node transformer/while/body/parallel_0/body/decoder/layer_0/self_attention/multihead_attention/dot_product_attention/MatMul_1 (defined at /usr/local/lib/python2.7/dist-packages/tensor2tensor/layers/common_attention.py:1540)  = BatchMatMul[T=DT_FLOAT, adj_x=false, adj_y=false, _device="/job:localhost/replica:0/task:0/device:CPU:0"](transformer/while/body/parallel_0/body/decoder/layer_0/self_attention/multihead_attention/dot_product_attention/Print_6, transformer/while/body/parallel_0/body/decoder/layer_0/self_attention/multihead_attention/dot_product_attention/Print_7)]]

Basically, adding logits to bias is causing logits to have an invalid shape. This sort of seems to be a caching issue, since the previous queries,keys, and values are not found on the second autoregressive iteration.

**Comment by mikeymezher** [comment]
Scratch the bottom part, I found that one of my modifications was causing the cache not to be passed. 

Still, the top part of this post is valid (not a bug, just unreachable code)
